### Exploration of results

Case studies of individual companies, to gain in depth-insights into model versus analyst reasoning.

In [1]:
import pandas as pd
import os

In [2]:
os.getcwd()

'c:\\Users\\benny\\OneDrive\\Studium\\Tübingen\\DS_in_B&E\\Masterarbeit\\code'

- Comprehensible list to see the specific timeframe in which a stock was part of the analysis

In [3]:
def extract_signal(text):
    # Normalize text: lowercase, remove punctuation, remove extra spaces
    text_clean = ''.join(c for c in text.lower())
    
    # Map strong signals to regular ones
    if 'strong buy' in text_clean:
        return 'buy'
    if 'strong sell' in text_clean:
        return 'sell'
    if 'buy' in text_clean:
        return 'buy'
    if 'sell' in text_clean:
        return 'sell'
    if 'hold' in text_clean:
        return 'hold'
    return None 

In [4]:
# List of companies to show when they were present in the dataset
ciks_and_dates = pd.read_csv("../results/shared_ciks_and_dates.csv", dtype ={"cik": str})
# Convert date column to datetime for analysis
ciks_and_dates["date"] = pd.to_datetime(ciks_and_dates["date"])

In [5]:
# For every CIK, find the first and last date they appear in the dataset
cik_date_ranges = ciks_and_dates.groupby("cik")["date"].agg(["min", "max"]).reset_index()
cik_date_ranges.columns = ["cik", "first_date", "last_date"]
cik_date_ranges["first_date"] = cik_date_ranges["first_date"].dt.strftime('%m-%Y')
cik_date_ranges["last_date"] = cik_date_ranges["last_date"].dt.strftime('%m-%Y')

In [6]:
stock_prices = pd.read_csv("../data/sp1500_monthly_prices.csv", dtype={"cik": str})
stock_prices["date"] = pd.to_datetime(stock_prices["date"])
stock_prices["date"] = stock_prices["date"].dt.strftime('%m-%Y')

In [7]:
cik_date_ranges = cik_date_ranges.merge(stock_prices[["cik", "security"]],on=["cik"], how="left").drop_duplicates(subset = ["cik"]).reset_index(drop=True)
cik_date_ranges.to_csv("../results/analysis_window_by_cik.csv", index=False)

- Simple check, if Analyst/LLM recommendations predicted a rising/falling stock price correctly i.e. buy/hold/sell was recommended

In [8]:
# Load in LLM and Analyst ratings and format date accordingly
llm_ratings = pd.read_csv("../results/llm_recommendations_final.csv", dtype={"cik": str}, low_memory=False)
llm_ratings["date"] = pd.to_datetime(llm_ratings["date"])
llm_ratings["date"] = llm_ratings["date"].dt.strftime('%m-%Y')
llm_ratings["action"] = llm_ratings["rating"].apply(extract_signal)
analyst_ratings = pd.read_csv("../results/sp1500_sell_side_recommendations_ffilled.csv", dtype={"cik": str}, low_memory=False).drop(columns=["mean_rating", "rating", "security"], axis = 1)
analyst_ratings["date"] = pd.to_datetime(analyst_ratings["date"])
analyst_ratings["date"] = analyst_ratings["date"].dt.strftime('%m-%Y')

In [9]:
llm_ratings

cik     date       rating action
0       0001497645  03-2010         sell   sell
1       0001497645  06-2010         sell   sell
2       0001497645  09-2010         sell   sell
3       0001497645  12-2010         sell   sell
4       0001497645  03-2011         sell   sell
...            ...      ...          ...    ...
147468  0001822993  12-2022  strong sell   sell
147469  0001822993  03-2023         sell   sell
147470  0001822993  06-2023          buy    buy
147471  0001822993  09-2023          buy    buy
147472  0001822993  12-2023         sell   sell

[147473 rows x 4 columns]

In [10]:
# Merge both with stock prices
llm_ratings = llm_ratings.merge(stock_prices, on=["cik", "date"], how="left").drop(columns = ["security"], axis = 1)
analyst_ratings = analyst_ratings.merge(stock_prices, on=["cik", "date"], how="left").drop(columns = ["security"], axis = 1)
# Drop NAs
llm_ratings = llm_ratings.dropna()
analyst_ratings = analyst_ratings.dropna()

In [11]:
# Determine which ciks are present in both datasets on same date
shared_ciks_and_dates = pd.merge(llm_ratings[["cik", "date"]], analyst_ratings[["cik", "date"]], on=["cik", "date"], how="inner").drop_duplicates().reset_index(drop=True)

In [12]:
# Only keep ratings for those ciks and dates
llm_ratings = llm_ratings.merge(shared_ciks_and_dates, on=["cik", "date"], how="inner").reset_index(drop=True).drop_duplicates(subset = ["cik", "date"]).reset_index(drop=True)
analyst_ratings = analyst_ratings.merge(shared_ciks_and_dates, on=["cik", "date"], how="inner").reset_index(drop=True).drop_duplicates(subset = ["cik", "date"]).reset_index(drop=True)

In [13]:
# Shift recommendationms by one month to align with future stock returns
llm_ratings["next_price"] = llm_ratings.groupby("cik")["price"].shift(-1)
analyst_ratings["next_price"] = analyst_ratings.groupby("cik")["price"].shift(-1)

In [14]:
def compute_actual_movement(row):
    change = (row['next_price'] - row['price']) / row['price']
    if abs(change) <= 0.05: return 'hold'
    return 'buy' if change > 0 else 'sell'

In [15]:
# Determine if recommendation was "correct"
# Logic: If signal was "buy" and the following price is higher, the recommendation was correct.
# If signal was "sell" and the following price is lower, the recommendation was correct.
# If signal was "hold" and the following price is within +-5% of the current price, the recommendation was correct.
llm_ratings["rec_correct"] = (
    ((llm_ratings["next_price"] > llm_ratings["price"]) & (llm_ratings["action"] == "buy")) | 
    ((llm_ratings["next_price"] < llm_ratings["price"]) & (llm_ratings["action"] == "sell")) | 
    (((llm_ratings["next_price"] - llm_ratings["price"]).abs() / llm_ratings["price"] <= 0.02) & (llm_ratings["action"] == "hold"))
)
# Add actual movement column
llm_ratings["actual_movement"] = llm_ratings.apply(compute_actual_movement, axis=1)
llm_ratings

cik     date       rating action  price  next_price  \
0       0001497645  03-2011         sell   sell   9.94       11.35   
1       0001497645  06-2011         sell   sell  11.35        7.06   
2       0001497645  09-2011         sell   sell   7.06        9.44   
3       0001497645  12-2011         sell   sell   9.44        7.58   
4       0001497645  03-2012         sell   sell   7.58        8.37   
...            ...      ...          ...    ...    ...         ...   
111459  0001822993  12-2022  strong sell   sell  34.79       37.41   
111460  0001822993  03-2023         sell   sell  37.41       30.61   
111461  0001822993  06-2023          buy    buy  30.61       38.22   
111462  0001822993  09-2023          buy    buy  38.22       51.20   
111463  0001822993  12-2023         sell   sell  51.20         NaN   

        rec_correct actual_movement  
0             False             buy  
1              True            sell  
2             False             buy  
3              True            sell  
4             False             buy  
...             ...             ...  
111459        False             buy  
111460         True            sell  
111461         True             buy  
111462         True             buy  
111463        False            sell  

[111464 rows x 8 columns]

Why next price missing fo 001497645 on 12-2023 since price for 03-2000 is present?

In [16]:
analyst_ratings ["rec_correct"] = (
    ((analyst_ratings["next_price"] > analyst_ratings["price"]) & (analyst_ratings["action"] == "buy")) |
    ((analyst_ratings["next_price"] < analyst_ratings["price"]) & (analyst_ratings["action"] == "sell")) | 
    (((analyst_ratings["next_price"] - analyst_ratings["price"]).abs() / analyst_ratings["price"] <= 0.02) & (analyst_ratings["action"] == "hold"))
)
# Add actual movement column
analyst_ratings["actual_movement"] = analyst_ratings.apply(compute_actual_movement, axis=1)
analyst_ratings

cik     date action    price  next_price  rec_correct  \
0       0000001750  02-2000    buy   23.750      13.875        False   
1       0000001750  05-2000    buy   13.875      11.250        False   
2       0000001750  08-2000    buy   11.250      10.375        False   
3       0000001750  11-2000    buy   10.375      13.600         True   
4       0000001750  02-2001    buy   13.600      14.000         True   
...            ...      ...    ...      ...         ...          ...   
111459  0002012383  12-2022    buy  708.630     669.120        False   
111460  0002012383  03-2023    buy  669.120     691.140         True   
111461  0002012383  06-2023    buy  691.140     646.490        False   
111462  0002012383  09-2023    buy  646.490     811.800         True   
111463  0002012383  12-2023    buy  811.800         NaN        False   

       actual_movement  
0                 sell  
1                 sell  
2                 sell  
3                  buy  
4                 hold  
...                ...  
111459            sell  
111460            hold  
111461            sell  
111462             buy  
111463            sell  

[111464 rows x 7 columns]

In [17]:
print(f"Accuracy of LLM recommendations: {llm_ratings['rec_correct'].mean()}")
print(f"Accuracy of Analyst recommendations: {analyst_ratings['rec_correct'].mean()}")

Accuracy of LLM recommendations: 0.47093231895499893
Accuracy of Analyst recommendations: 0.4320856958300438


In [18]:
from sklearn.metrics import classification_report
def get_formatted_report(df):

    # Copy of df with no NAs in actual_movement or action
    eval_df = df.dropna(subset=['actual_movement', 'action']).copy()
    
    # Report as dictionary, so that it can easier be manipulated
    report_dict = classification_report(
        eval_df['actual_movement'], 
        eval_df['action'], 
        output_dict=True
    )
    
    # No averages needed
    keys_to_remove = ['accuracy', 'macro avg', 'weighted avg']
    for key in keys_to_remove:
        report_dict.pop(key, None)
        
    # Back to DataFrame for better formatting
    report_df = pd.DataFrame(report_dict).transpose()
    # Capitalize everything
    report_df.index = [str(i).capitalize() for i in report_df.index]
    report_df.columns = [str(i).capitalize() for i in report_df.columns]
    # Drop number of samples column and rounding
    report_df = report_df.drop(columns=['Support'], axis=1)
    report_df = report_df.round(4)
    
    return report_df

In [19]:
llm_classification_report = get_formatted_report(llm_ratings)
llm_classification_report.to_csv("../results/llm_classification_report.csv")
llm_classification_report

Precision  Recall  F1-score
Buy      0.4665  0.4418    0.4538
Hold     0.2868  0.1406    0.1886
Sell     0.3340  0.5131    0.4047

In [20]:
analyst_classification_report = get_formatted_report(analyst_ratings)
analyst_classification_report.to_csv("../results/analyst_classification_report.csv")
analyst_classification_report

Precision  Recall  F1-score
Buy      0.4431  0.6847     0.538
Hold     0.2830  0.3355     0.307
Sell     0.3074  0.0087     0.017